File: Data Structure Lab/MergeSort.cpp
#include<stdio.h>
void Merge(int *a,int lb,int mid,int ub)
{
	int i,j,k=0,b[10];
	i=lb;
	j=mid+1;
	k=lb;
	while(i<=mid && j<=ub)
	{
		if(a[i]<a[j])
		{
			b[k]=a[i];
			i++;k++;
		}
		else
		{
			b[k]=a[j];
			j++;k++;
		}
	}
	if(i>mid)
	{
		while(j<=ub)
		{
			b[k]=a[j];
			j++;k++;
		}
	}
	else
	{
		while(i<=mid)
		{
			b[k]=a[i];
			i++;k++;
		}
	}
	for(i=lb;i<=ub;i++)
	a[i]=b[i];
	
}

void MergeSort(int *a,int lb,int ub)
{
	if (lb<ub)
	{
		int mid=(lb+ub)/2;
		MergeSort(a,lb,mid);
		MergeSort(a,mid+1,ub);
		Merge(a,lb,mid,ub);
	}
}

int main()
{
	int a[10],i,n;
	printf("Enter size of array ");
	scanf("%d",&n);
	for(i=0;i<n;i++)
	scanf("%d",&a[i]);
	MergeSort(a,0,n-1);
	for(i=0;i<n;i++)
	printf("%d ",a[i]);
}

File: Lab/Python Lab/Lab8/Merge_Sort.py
def merge(l,lb,mid,ub):
    i=lb
    j=mid+1
    a=[0]*(ub+1)
    k=0
    while(i<=mid and j<=ub):
        if(l[i]<l[j]):
            a[k]=l[i]
            i+=1
        else:
            a[k]=l[j]
            j+=1
        k+=1
    while(i<=mid):
        a[k]=l[i]
        i+=1
        k+=1
    while(j<=ub):
        a[k]=l[j]
        j+=1
        k+=1

    k=0
    i=lb
    while(i<=ub):
        l[i]=a[k]
        i+=1
        k+=1
        


def merge_sort(l,lb,ub):
    if(lb<ub):
        mid=(lb+ub)//2
        merge_sort(l,lb,mid)
        merge_sort(l,mid+1,ub)
        merge(l,lb,mid,ub)


l=list(map(int,input("Enter Numbers: ").split()))
merge_sort(l,0,len(l)-1)
print(l)
